webserver/webserver/src/main/java/io/helidon/webserver/ReferenceHoldingQueue.java

```diff
@@ -122,6 +119,40 @@ private void unlink(ReleasableReference<T> reference) {
         set.remove(reference);
     }
 
+    /**
+     * An indirect reference is a phantom reference that also points to
+     * another object. When this reference is added to a {@code ReferenceQueue}
+     * after its reference becomes unreachable, it provides access to that
+     * other object via its {@code acquire} method. Note that the object can only
+     * be acquired once, its reference is set to null afterwards to ensure
+     * collection.
+     *
+     * @param <T> type of referent object
+     * @param <R> type of the other object
+     */
+    static class IndirectReference<T, R> extends PhantomReference<T> {
+        private final AtomicReference<R> otherRef = new AtomicReference<>();
+
+        IndirectReference(T referent, ReferenceQueue<? super T> q, R otherRef) {
+            super(referent, q);
+            this.otherRef.lazySet(otherRef);
```

Comment:
Why `lazySet`?

Reply:
then it is a normal `store` (no expensive full barrier)

(think `mov mm, r` vs `lock: mov mm, r`)

Reply:
Actually, I am not sure that the atomicity of the reference is even required here.

All accesses are meant to be single-threaded in the intended use. It is atomic only for the purpose of potential other uses.

Reply:
Maybe not needed, but seems like the safer option.